From: Jan Luebbe <jlu@pengutronix.de>
Date: Sat, 31 Jul 2021 19:09:15 +0200
Subject: [PATCH] lxatac: label ADC inputs and map some via
 iio-rescale/iio-hwmon

Also change vdd-supply from &vdda to &vdd. No other DTS uses vdda for
vdd and it doesn't seem to be connected that way.

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>
---
 arch/arm/boot/dts/stm32mp157c-lxa-tac.dts | 122 +++++++++++++++++++++++++++++-
 1 file changed, 121 insertions(+), 1 deletion(-)

diff --git a/arch/arm/boot/dts/stm32mp157c-lxa-tac.dts b/arch/arm/boot/dts/stm32mp157c-lxa-tac.dts
index d6dfca33b67f..5afcdc628595 100644
--- a/arch/arm/boot/dts/stm32mp157c-lxa-tac.dts
+++ b/arch/arm/boot/dts/stm32mp157c-lxa-tac.dts
@@ -81,6 +81,74 @@ reg_1v2: regulator_1v2 {
 		regulator-always-on;
 	};
 
+
+	iobus_curr: current-sense-amplifier@0 {
+		compatible = "current-sense-amplifier";
+		io-channels = <&adc1 5>;
+		#io-channel-cells = <0>;
+		label = "iobus-curr";
+		sense-resistor-micro-ohms = <2200000000>; /* 2.2k */
+		sense-gain-mult = <653>;
+		sense-gain-div = <1000000>;
+	};
+
+	iobus_volt: voltage-divider@0 {
+		compatible = "voltage-divider";
+		io-channels = <&adc1 9>;
+		#io-channel-cells = <0>;
+		label = "iobus-volt";
+		output-ohms = <1000>;
+		full-ohms = <6100>;
+	};
+
+	iobus {
+		compatible = "iio-hwmon";
+		io-channels = <&iobus_volt>, <&iobus_curr>;
+	};
+
+	host_curr: current-sense-amplifier@1 {
+		compatible = "current-sense-amplifier";
+		io-channels = <&adc1 13>;
+		#io-channel-cells = <0>;
+		label = "host-curr";
+		sense-resistor-micro-ohms = <1300000000>; /* 1.3k */
+		sense-gain-mult = <653>;
+		sense-gain-div = <1000000>;
+	};
+
+	host_1_curr: current-sense-amplifier@2 {
+		compatible = "current-sense-amplifier";
+		io-channels = <&adc1 15>;
+		#io-channel-cells = <0>;
+		label = "host-1-curr";
+		sense-resistor-micro-ohms = <1800000000>; /* 1.8k */
+		sense-gain-mult = <653>;
+		sense-gain-div = <1000000>;
+	};
+
+	host_2_curr: current-sense-amplifier@3 {
+		compatible = "current-sense-amplifier";
+		io-channels = <&adc1 0>; /* via ANA0 */
+		#io-channel-cells = <0>;
+		label = "host-2-curr";
+		sense-resistor-micro-ohms = <1800000000>; /* 1.8k */
+		sense-gain-mult = <653>;
+		sense-gain-div = <1000000>;
+	};
+
+	host_3_curr: current-sense-amplifier@4 {
+		compatible = "current-sense-amplifier";
+		io-channels = <&adc1 1>; /* via ANA1 */
+		#io-channel-cells = <0>;
+		label = "host-3-curr";
+		sense-resistor-micro-ohms = <1800000000>; /* 1.8k */
+		sense-gain-mult = <653>;
+		sense-gain-div = <1000000>;
+	};
+	usb {
+		compatible = "iio-hwmon";
+		io-channels = <&host_curr>, <&host_1_curr>, <&host_2_curr>, <&host_3_curr>;
+	};
 };
 
 /* VREFBUF: What about the external reference voltage? */
@@ -94,7 +162,7 @@ &vrefbuf {
 &adc {
 	pinctrl-names = "default";
 	pinctrl-0 = <&adc1_ain_pins_a>;
-	vdd-supply = <&vdda>;
+	vdd-supply = <&vdd>;
 	vdda-supply = <&vdda>;
 	vref-supply = <&vrefbuf>;
 	status = "okay";
@@ -102,11 +170,63 @@ adc1: adc@0 {
 		st,adc-channels = <0 1 2 5 6 9 10 11 13 14 15 18>;
 		st,min-sample-time-nsecs = <5000>;
 		status = "okay";
+		channel@0 {
+			reg = <0>;
+			label = "HOST_2_CURR_FB";
+		};
+		channel@1 {
+			reg = <1>;
+			label = "HOST_3_CURR_FB";
+		};
+		channel@2 {
+			reg = <2>;
+			label = "OUT_0_FB";
+		};
+		channel@5 {
+			reg = <5>;
+			label = "IOBUS_CURR_FB";
+		};
+		channel@6 {
+			reg = <6>;
+			label = "PWR_U_VCC"; /* via STACK_SPI_CS0 to Power Switch Board */
+		};
+		channel@9 {
+			reg = <9>;
+			label = "IOBUS_VOLT_FB";
+		};
+		channel10 {
+			reg = <10>;
+			label = "OUT_1_FB";
+		};
+		channel@11 {
+			reg = <11>;
+			label = "PWR_U_GND"; /* via STACK_SPI_CS1 to Power Switch Board */
+		};
+		channel@13 {
+			reg = <13>;
+			label = "HOST_CURR_FB";
+		};
+		channel@14 {
+			reg = <14>;
+			label = "PWR_I"; /* via STACK_SPI_CS2 to Power Switch Board */
+		};
+		channel@15 {
+			reg = <15>;
+			label = "HOST_1_CURR_FB";
+		};
+		channel@18 {
+			reg = <18>;
+			label = "STACK_SPI_CS3"; /* FIXME: unused? */
+		};
 	};
 	adc2: adc@100 {
 		st,adc-channels = <12>;
 		st,min-sample-time-nsecs = <500000>;
 		status = "okay";
+		channel@12 {
+			reg = <12>;
+			label = "internal temperature";
+		};
 	};
 };
 
